oeps/oep-0005.rst

```diff
@@ -0,0 +1,182 @@
+   makes it hard to update (and to push updates out to users).
+
+
+Specification
+=============
+
+Interface
+~~~~~~~~~
+
+The primary developer interaction with ``devstack`` will be through ``make``
+shortcuts. The following targets will be standard::
+
+    make devstack.watch  # Start the devstack services (if they aren't already)
+                         # and display their logs on stdout
+
+    make devstack.ssh    # Start the devstack services (if they aren't already)
+                         # and ssh into the primary service (as the user that
+                         # owns the service, in the directory that has the
+                         # service source code)
+
+    make devstack.clean  # Stop the devstack services and remove any intermediate
+                         # data that they stored in the .devstack/ directory
+
+    make devstack.*      # Start the devstack services (if they aren't already)
+                         # and run `make *` from the code directory of the service
+                         # (as the service owner)
+
+Implementation
+~~~~~~~~~~~~~~
+
+Requirements
+------------
+
+1. Services that aren't part of the Open edX ecosystem must be run using
+   their own published containers (where available).
+2. Open edX containers must run all of their processes via `supervisor`_ to
+   provide a single entrypoint even when the container has multiple processes
+   (such as `nginx`_ and a `Django`_ service running under `gunicorn`_).
+3. Open edX services running in containers must log to ``stdout`` and ``stderr``
+   so that those logs can be captured by `supervisor`_.
+4. Dependent services (such as databases) should be configured to mount their
+   storage directories in ``.devstack/`` inside the source code directory (this
+   should be added to `.gitignore` and `.dockerignore`_. This will allow data from
+   ``devstack`` to be persist across container executions.
+5. The Open edX service must mount the source code directory from the host
+   machine into the ``devstack`` container so that ``devstack`` is always running
+   the version of code being edited.
+6. ``devstack`` services should be configured to auto-reload when source code
+   is modified.
+7. Open edX Service containers should be built during CI, and published so that any
+   developer can download the latest container. They should also be tagged for
+   each versioned release.
+
+Technology Selection
+--------------------
+
+The prototype implementation was done using `Docker`_ for the containers, and
+`Docker Compose`_ to manage the networking between services.
+
+Rationale
+=========
+
+Choice of `Docker`_ over `Vagrant`_
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+While `Vagrant`_ has worked relatively well for the current monolithic Devstack
+implementation, it would be too resource heavy for running many services simultaneously.
+`Docker`_ has a much lower per-service overhead.
```

Comment:
I'm not sure that converting to docker has a significantly higher development time than converting to vagrant multi-machine would. The biggest difficulties are just in the setup of devstack to run as multiple services with network boundaries, rather than docker specific issues (I believe).

As far as resource usage goes, I think the advantage that Docker has in a development environment is that, unlike a pure virtualization solution like virtualbox, Docker *doesn't* have resource allocation guarantees, which means that the all of the services can share the resources on the host box, rather than having to explicitly parcel them out to each of the VMs. That said, I haven't done an apples-to-apples comparison on an edx service in particular to see how the strategies compare.